articles/service-fabric/service-fabric-work-with-reliable-collections.md

```diff
@@ -213,10 +213,53 @@ Internally, Reliable Collections serialize your objects using .NET's DataContrac
 Furthermore, service code is upgraded one upgrade domain at a time. So, during an upgrade, you have two different versions of your service code running simultaneously. You must avoid having the new version of your service code use the new schema as old versions of your service code might not be able to handle the new schema. When possible, you should design each version of your service to be forward compatible by one version. Specifically, this means that V1 of your service code should be able to ignore any schema elements it does not explicitly handle. However, it must be able to save any data it doesn't explicitly know about and write it back out when updating a dictionary key or value.
 
 > [!WARNING]
-> While you can modify the schema of a key, you must ensure that your key's hash code and equals algorithms are stable. If you change how either of these algorithms operate, you will not be able to look up the key within the reliable dictionary ever again.
+> While you can modify the schema of a key, you must ensure that your key's equality and comparison algorithms are stable.
+> Behavior of reliable collections after a change in either of these algorithms is undefined and may lead to data corruption, loss and
+> service crashes.
 > .NET Strings can be used as a key but use the string itself as the key--do not use the result of String.GetHashCode as the key.
 
-Alternatively, you can perform what is typically referred to as a two upgrade. With a two-phase upgrade, you upgrade your service from V1 to V2: V2 contains the code that knows how to deal with the new schema change but this code doesn't execute. When the V2 code reads V1 data, it operates on it and writes V1 data. Then, after the upgrade is complete across all upgrade domains, you can somehow signal to the running V2 instances that the upgrade is complete. (One way to signal this is to roll out a configuration upgrade; this is what makes this a two-phase upgrade.) Now, the V2 instances can read V1 data, convert it to V2 data, operate on it, and write it out as V2 data. When other instances read V2 data, they do not need to convert it, they just operate on it, and write out V2 data.
+Alternatively, you can perform a multi-phase upgrade. 
+1. Upgrade service to a new version that
+    - has both the original V1, and the new V2 version of the data contracts included in the service code package;
+    - registers custom V2 [state serializers](/azure/service-fabric/service-fabric-reliable-services-reliable-collections-serialization#custom-serialization), if needed;
+    - performs all operations on the original, V1 collection using the V1 data contracts.
+2. Upgrade service to a new version that
+    - [creates a new, V2 collection](/dotnet/api/microsoft.servicefabric.data.ireliablestatemanager.getoraddasync);
+    - performs each add, update and delete operation on first V1 and then V2 collections in a single transaction;

```

Comment:
While it might be straightfoward to me, I think it might be good to point out to the customer that certain update/delete operation for V2 collection will throw exception because it does not exist. And, that can be safely ignored since it will be handled in step 3.